### PR TITLE
UCT/UD/MLX5: Make support of zero copy APIs in ud_mlx5

### PR DESCRIPTION
**Pull request by shssf**
Related Issue #951

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2582/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1938/ for details (Mellanox internal link).

File: src/uct/ib/ud/accel/ud_mlx5.c [hunk 1]
```diff
@@ -644,7 +647,9 @@ static UCS_CLASS_INIT_FUNC(uct_ud_mlx5_iface_t,
     UCS_CLASS_CALL_SUPER_INIT(uct_ud_iface_t, &uct_ud_mlx5_iface_ops,
                               md, worker, params, 0, config);
 
-    uct_ib_iface_set_max_iov(&self->super.super, 1);
+    uct_ib_iface_set_max_iov(&self->super.super,
+                             (UCT_UD_MLX5_WQE_SIZE - sizeof(struct mlx5_wqe_ctrl_seg)) /
```
Comment (yosefe) [inline src/uct/ib/ud/accel/ud_mlx5.c]: UCT_UD_MLX5_WQE_SIZE is only the size of ctrl+address, it is not the maximal possible size
the maximal size is (UCT_IB_MLX5_MAX_BB * MLX5_SEND_WQE_BB). and need to do something like:
((UCT_IB_MLX5_MAX_BB * MLX5_SEND_WQE_BB) - sizeof(ctrl)-sizeof(dgram)-sizeof(inl) -max_hdr) / sizeof(dseg)
(i noticed that max_inline was also calculated incorrectly in ud_mlx5)

File: src/uct/ib/ud/accel/ud_mlx5.c [hunk 2]
```diff
@@ -659,7 +650,13 @@ static UCS_CLASS_INIT_FUNC(uct_ud_mlx5_iface_t,
...
                               md, worker, params, 0, &config->super);
 
-    uct_ib_iface_set_max_iov(&self->super.super, 1);
+    uct_ib_iface_set_max_iov(&self->super.super,
+                             ((UCT_IB_MLX5_MAX_BB * MLX5_SEND_WQE_BB) -
+                             sizeof(struct mlx5_wqe_ctrl_seg) -
+                             UCT_IB_MLX5_AV_FULL_SIZE -
+                             sizeof(struct mlx5_wqe_inl_data_seg) -
+                             self->super.config.max_inline) /
+                             sizeof(struct mlx5_wqe_data_seg));
```
Comment (shssf) [inline src/uct/ib/ud/accel/ud_mlx5.c]: @yosefe FYI. Current equation provide `max_iov == 4`.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2595/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2596/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1953/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2598/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1955/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1956/ for details (Mellanox internal link).

File: src/uct/ib/ud/accel/ud_mlx5.c [hunk 3]
```diff
@@ -659,7 +649,8 @@ static UCS_CLASS_INIT_FUNC(uct_ud_mlx5_iface_t,
     UCS_CLASS_CALL_SUPER_INIT(uct_ud_iface_t, &uct_ud_mlx5_iface_ops,
                               md, worker, params, 0, &config->super);
 
-    uct_ib_iface_set_max_iov(&self->super.super, 1);
+    uct_ib_iface_set_max_iov(&self->super.super, UCT_IB_MLX5_AM_ZCOPY_MAX_IOV);
+    self->super.config.max_inline = UCT_IB_MLX5_AM_MAX_HDR(UCT_IB_MLX5_AV_FULL_SIZE);
```
Comment (shssf) [inline src/uct/ib/ud/accel/ud_mlx5.c]: `max_inline` length fixed (was 124, current 140)

File: src/uct/ib/mlx5/ib_mlx5.h [hunk 4]
```diff
@@ -62,6 +62,15 @@ struct mlx5_grh_av {
 #  define MLX5_WQE_CTRL_SOLICITED  (1<<1)
 #endif
 
+#define UCT_IB_MLX5_AM_ZCOPY_MAX_IOV  3UL
+
+#define UCT_IB_MLX5_AM_MAX_HDR(_av_size) \
```
Comment (shssf) [inline src/uct/ib/mlx5/ib_mlx5.h]: The macro `UCT_RC_MLX5_AM_MAX_HDR` moved into common place and reused in UD.
The name changed to `UCT_IB_MLX5_AM_MAX_HDR`.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2601/ for details.

File: src/uct/ib/base/ib_iface.h [hunk 5]
```diff
@@ -354,7 +354,12 @@ size_t uct_ib_iface_get_max_iov(uct_ib_iface_t *iface)
 static UCS_F_ALWAYS_INLINE
 void uct_ib_iface_set_max_iov(uct_ib_iface_t *iface, size_t max_iov)
 {
-    iface->config.max_iov = ucs_min(UCT_IB_MAX_IOV, max_iov);
+    size_t min_iov_requested;
+
+    ucs_assert((ssize_t)max_iov > 0);
```
Comment (shssf) [inline src/uct/ib/base/ib_iface.h]: avoid unexpected big `max_iov` by `ucs_assert`.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1959/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2609/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/4/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2640/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/38/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1992/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1994/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
bot:bgateNEW:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/71/ for details.

**Pull request merged**